### ISSUE TITLE
DSG pages crash when called because a path cannot be resolved (Cannot find module)

### ISSUE THREAD

**Issue by timidak** [gatsbyjs/gatsby#38265]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

In our project we use the gatsby-resolve-src plugin which is equivalent to adjusting the `webpack.config` to resolve modules from the src directory. During the build everything works fine and we can visit SSG pages after using `gatsby serve`. This also works on Netlify.
But when it comes to DSG sites, we have the problem that visiting them results in an error. 

```shell
Rendering html for "/electronics" failed. Cannot find module 'constants_new'

   1 | query ProductsQuery {
   2 |   allProduct {
   3 |     nodes {
   4 |       id
   5 |       price
   6 |       title
   7 |       productId
>  8 |       priceInEuro
     |       ^
   9 |     }
  10 |   }
  11 | }



  Error: Cannot find module 'constants_new'

  - 2.index.js:13 webpackMissingModule
   /experiments/gatsby-ssr-exp/.cache/query-engine/2.index.js:13:50

  - 2.index.js:13 Module.2357
   /experiments/gatsby-ssr-exp/.cache/query-engine/2.index.js:13:138

  - index.js:310118 Function.__webpack_require__
    /experiments/gatsby-ssr-exp/.cache/query-engine/index.js:310118:42

  - task_queues:95 processTicksAndRejections
    node:internal/process/task_queues:95:5

  - index.js:310039 resolve
   /experiments/gatsby-ssr-exp/.cache/query-engine/index.js:310039:30
```

### For context

In our project, we use API calls and other functions ("controllers") to enhance the data model. I reproduced this behavior in my reproduction, by calling an api and importing a controller which then imports a constant.

It seem's that in the DSG context (and probably also in the SSR context) gatsby cannot resolve the paths. I think the error could be that the webpack config is not working properly in this context or imports in the resolvers are not processed properly by webpack. 

Using relative paths would fix the bug, however our codebase is so large that this would require significant effort. Furthermore, in my opinion, it seems to be a bug that might manifest itself differently later on.

### Reproduction Link

https://github.com/timidak/exp-gatsby-ssr/

### Steps to Reproduce

1. Clone repo
2. Install node_modules
3. Run `npm run build`
4. Run `npm run serve`
5. Visit localhost:9000 (works ✅ )
6. Visit localhost:9000/electronics (a category site) 
7. Gatsby crashes with `Internal Server Error`
8. This is also reproducable on Netlify
...

### Expected Result

The DSG site should render properly and the paths should resolve, because it is no option to refactor all parts of our big code base back to relative paths. 

### Actual Result

The site crashed when opening an DSG site. With the above mentioned error.

### Environment

```shell
System:
    OS: macOS 12.6
    CPU: (10) arm64 Apple M1 Max
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.12.1 - 
    Yarn: 1.22.19 - 
    npm: 8.19.2 - 
  Browsers:
    Chrome: 114.0.5735.133
    Firefox: 111.0.1
    Safari: 16.0
  npmPackages:
    gatsby: ^5.10.0 => 5.10.0 
    gatsby-plugin-netlify: ^5.1.0 => 5.1.0 
    gatsby-plugin-resolve-src: ^2.1.0 => 2.1.0
```

**Comment by LekoArts** [comment]
Hello!

Unfortunately this is a current limitation and as such I must treat this as a feature request. I've updated the docs to outline that limitation: https://github.com/gatsbyjs/gatsby/commit/0c8f948af935353dd0e774f555052d18b5fd8fba
You can open a feature request here: https://github.com/gatsbyjs/gatsby/discussions/categories/ideas-feature-requests

To explain it in more detail:
We use webpack to bundle everything for DSG to run. `onCreateWebpackConfig` doesn't change that webpack configuration though, it only changes our "main" configuration. As such, the webpack plugin you're using isn't changing anything for DSG and inside your `createResolvers` the path aliasing won't work.

So in a nutshell, you won't need to change all imports, you need to use relative paths in files that are imported inside `gatsby-node` so that the DSG engine can resolve everything.
